scipy/_lib/_array_api.py

```diff
@@ -172,9 +174,14 @@ def _asarray(
     `subok` is included to allow this function to preserve the behaviour of
     `np.asanyarray` for NumPy based inputs.
     """
-    if xp is None:
-        xp = array_namespace(array)
-    if is_numpy(xp):
+    if xp is not None:
+        assert xp_out is xp_in is None
+        xp_out = xp_in = xp
+    
+    xp_out = array_namespace(array) if xp_out is None else xp_out
+    xp_in = array_namespace(array) if xp_in is None else xp_in
+
+    if is_numpy(xp_out) and is_numpy(xp_in):
         # Use NumPy API to support order
         if copy is True:
             array = np.array(array, order=order, dtype=dtype, subok=subok)
```

Comment:
I'm wondering whether we should do something when this function is called with options that aren't supported by the array library? https://github.com/scipy/scipy/pull/18668/files#r1841587120